packages/transaction-controller/src/TransactionController.ts

```diff
@@ -443,6 +452,27 @@ export class TransactionController extends BaseController<
     const isEIP1559Compatible = await this.getEIP1559Compatibility();
     validateTxParams(txParams, isEIP1559Compatible);
 
+    if (origin === ORIGIN_METAMASK) {
+      // Ensure the 'from' address matches the currently selected address
+      const selectedAddress = this.getSelectedAddress();
+      if (txParams.from !== selectedAddress) {
+        throw rpcErrors.internal({
+          message: `Internally initiated transaction is using invalid account.`,
+          data: {
+            origin,
+            fromAddress: txParams.from,
+            selectedAddress,
+          },
+        });
+      }
+    } else {
+      // Check if the origin has permissions to initiate transactions from the specified address
+      const permittedAddresses = await this.getPermittedAccounts(origin);
+      if (!permittedAddresses.includes(txParams.from)) {
+        throw providerErrors.unauthorized({ data: { origin: origin ?? '' } });
```

Comment:
Both client middlewares always pass the origin from the request, but as we have it defined as optional here, should we instead skip the entire validation if there is no origin otherwise it doesn't make sense to call `getPermittedAccounts` at all?